ffi/derive/src/arg.rs

```diff
@@ -0,0 +1,522 @@
+                    }
+
+                    self.visit_item_binding(last_seg);
+                    let binding_ffi_type = &self.ffi_type;
+                    let binding_src_to_ffi = &self.src_to_ffi;
+                    let binding_ffi_to_src = &self.ffi_to_src;
+
+                    let slice_len_arg_name = crate::bindgen::gen_slice_len_arg_name(arg_name);
+                    self.ffi_type = Some(parse_quote! {*mut #binding_ffi_type});
+                    self.src_to_ffi = Some(parse_quote! {
+                        let #arg_name = #arg_name.into_iter().map(|arg| {
+                            #binding_src_to_ffi
+                            arg
+                        });
+                    });
+                    self.ffi_to_src = Some(parse_quote! {
+                        let #arg_name = core::slice::from_raw_parts_mut(#arg_name, #slice_len_arg_name)
+                            .into_iter().map(|&mut arg| {
+                                #binding_ffi_to_src
+                                arg
+                            });
+                    });
+                }
+                "Into" => {
+                    self.visit_type(generic_arg_types(last_seg)[0]);
+                }
+                _ => abort!(trait_, "Unsupported `impl trait`"),
+            }
+        }
+    }
+    fn visit_type_infer(&mut self, _: &'ast syn::TypeInfer) {
+        unreachable!("Infer type not possible in a declaration")
+    }
+    fn visit_type_macro(&mut self, _: &'ast syn::TypeMacro) {
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_never(&mut self, _: &'ast syn::TypeNever) {
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_param(&mut self, _: &'ast syn::TypeParam) {
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_param_bound(&mut self, _: &'ast syn::TypeParamBound) {
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_paren(&mut self, _: &'ast syn::TypeParen) {
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_path(&mut self, node: &'ast syn::TypePath) {
+        let last_seg = node.path.segments.last().expect_or_abort("Defined");
+
+        let arg_name = self.name;
+        match last_seg.ident.to_string().as_str() {
+            "bool" => {
+                if self.is_input {
+                    self.ffi_type = Some(parse_quote! { u8 });
+                    self.src_to_ffi = Some(parse_quote! {let #arg_name = #arg_name as u8;});
+                    self.ffi_to_src = Some(parse_quote! {let #arg_name = #arg_name != 0;});
+                } else {
+                    self.ffi_type = Some(node.clone().into());
+                    self.src_to_ffi = Some(parse_quote! {();});
+                    self.ffi_to_src = Some(parse_quote! {();});
+                }
+            }
+            "u8" | "u16" => {
+                self.ffi_type = Some(parse_quote! {u32});
```

Comment:
because wasm doesn't supports only `u32`/`i32` and `f32`/`f64`. In the future we can introduce a flag to distinguish between wasm FFI and the usually supported FFI types